contracts/registrar/ZNSRootRegistrar.sol

```diff
@@ -77,59 +77,82 @@ contract ZNSRootRegistrar is
         // Confirms string values are only [a-z0-9-]
-        name.validate();
+        registration.name.validate();
 
         // Create hash for given domain name
-        bytes32 domainHash = keccak256(bytes(name));
+        bytes32 domainHash = keccak256(bytes(registration.name));
 
         if (registry.exists(domainHash))
             revert DomainAlreadyExists(domainHash);
 
         // Get price for the domain
-        uint256 domainPrice = rootPricer.getPrice(0x0, name, true);
+        uint256 domainPrice = rootPricer.getPrice(0x0, registration.name, true);
 
         _coreRegister(
             CoreRegisterArgs(
                 bytes32(0),
                 domainHash,
                 msg.sender,
-                domainAddress,
+                registration.domainAddress,
                 domainPrice,
                 0,
-                name,
-                tokenURI,
+                registration.name,
+                registration.tokenURI,
                 true,
-                paymentConfig
+                registration.paymentConfig
             )
         );
 
-        if (address(distributionConfig.pricerContract) != address(0)) {
+        if (address(registration.distributionConfig.pricerContract) != address(0)) {
             // this adds additional gas to the register tx if passed
-            subRegistrar.setDistributionConfigForDomain(domainHash, distributionConfig);
+            subRegistrar.setDistributionConfigForDomain(domainHash, registration.distributionConfig);
         }
 
         return domainHash;
     }
 
+    /**
+     * @notice This function allows registering multiple root domains in a single transaction.
+     * It iterates through an array of `RootDomainRegistrationParams` objects, registering each domain
+     * by calling the `registerRootDomain` function for each entry.
+     * @dev This function reduces the number of transactions required to register multiple domains,
+     * saving gas and improving efficiency. Each domain registration is processed sequentially.
+     * @param registrations An array of `RootDomainRegistrationParams` structs, each containing:
+     *      + `name`: The name (label) of the domain to register.
+     *      + `domainAddress`: The address to associate with the domain in the resolver (optional).
+     *      + `tokenURI`: The URI to assign to the domain token.
+     *      + `distributionConfig`: The distribution configuration for the domain (optional).
+     *      + `paymentConfig`: The payment configuration for the domain (optional).
+     * @return domainHashes An array of `bytes32` hashes representing the registered domains.
+     */
+    function registerMultipleRootDomains(
+        RootDomainRegistrationParams[] calldata registrations
+    ) external override returns (bytes32[] memory) {
+        bytes32[] memory domainHashes = new bytes32[](registrations.length);
+
+        for (uint256 i = 0; i < registrations.length; i++) {
```

Comment:
minor gas improvement is made by incrementing the counter outside of the loop in an unchecked block, and for such a heavy operation that will matter a lot

```solidity
uint256 i = 0;
for(i; i < registrations.length;) {
  ...
  unchecked {
    ++i;  
  }
}
```